sshd-common/src/main/java/org/apache/sshd/common/cipher/BaseAEADCipher.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements. See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership. The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.sshd.common.cipher;
+
+import org.apache.sshd.common.util.security.SecurityUtils;
+
+import javax.crypto.Cipher;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+import java.security.InvalidAlgorithmParameterException;
+import java.security.InvalidKeyException;
+import java.security.spec.AlgorithmParameterSpec;
+
+public abstract class BaseAEADCipher extends BaseCipher {
+
+    private Mode mode;
+    private SecretKey secretKey;
+    private AlgorithmParameterSpec params;
+    private boolean aadWritten;
+
+    public BaseAEADCipher(
+            int ivsize, int authSize, int kdfSize, String algorithm, int keySize, String transformation, int blkSize) {
+        super(ivsize, authSize, kdfSize, algorithm, keySize, transformation, blkSize);
+    }
+
+    @Override
+    protected byte[] initializeIVData(final Mode mode, final byte[] iv, final int reqLen) {
+        byte[] data = super.initializeIVData(mode, iv, reqLen);
+        params = initializeAlgorithmParameters(data);
+        return data;
+    }
+
+    @Override
+    protected byte[] initializeKeyData(final Mode mode, final byte[] key, final int reqLen) {
+        byte[] keyData = super.initializeKeyData(mode, key, reqLen);
+        secretKey = new SecretKeySpec(keyData, getAlgorithm());
+        return keyData;
+    }
+
+    @Override
+    protected Cipher createCipherInstance(final Mode mode, final byte[] key, final byte[] iv) throws Exception {
+        this.mode = mode;
+        Cipher instance = SecurityUtils.getCipher(getTransformation());
+        init(instance);
+        return instance;
+    }
+
+    private void init(Cipher cipher) throws InvalidKeyException, InvalidAlgorithmParameterException {
```

Comment:
Let's make it `protected` in the spirit of open source framework we try and provide our users the greatest possible flexibility in using and modifying our code to their needs. Thus very few pieces of code are truly private.

Reply:
Probably makes sense. I was trying to refactor this class a little so that supporting ChaCha20-Poly1305 later would be simpler since it's another AEAD algorithm, though it uses a slightly different key derivation algorithm in order to encrypt the packet length as well (wonder if an update or extension to AES/GCM can be made to do the same eventually).

Reply:
Seems I was a little off. The default ciphers in OpenSSH right now are `chacha20-poly1305@openssh.com,aes128-ctr,aes192-ctr,aes256-ctr,aes128-gcm@openssh.com,aes256-gcm@openssh.com`